src/storage/txn/store.rs

```diff
@@ -60,21 +58,30 @@ impl<S: Snapshot> SnapshotStore<S> {
         keys: &[Key],
         statistics: &mut Statistics,
     ) -> Result<Vec<Result<Option<Value>>>> {
-        // TODO: sort the keys and use ScanMode::Forward
-        let mut reader = MvccReader::new(
-            self.snapshot.clone(),
-            None,
-            self.fill_cache,
-            None,
-            None,
-            self.isolation_level,
-        );
-        let mut results = Vec::with_capacity(keys.len());
-        for k in keys {
-            results.push(reader.get(k, self.start_ts).map_err(Error::from));
+        use std::mem;
+
+        let mut order_and_keys: Vec<_> = keys.iter().enumerate().collect();
+        order_and_keys.sort_unstable_by(|(_, a), (_, b)| a.cmp(b));
+
+        let mut point_getter = PointGetterBuilder::new(self.snapshot.clone(), self.start_ts)
+            .fill_cache(self.fill_cache)
+            .isolation_level(self.isolation_level)
+            .build()?;
+
+        let mut values = Vec::with_capacity(keys.len());
+        unsafe {
```

Comment:
I recommend using `MaybeUninit` instead.

Reply:
Using uninitialized memory here is used based on the Vec's guarantee:

> If a Vec has allocated memory, then the memory it points to is on the heap (as defined by the allocator Rust is configured to use by default), and its pointer points to len initialized, contiguous elements in order (what you would see if you coerced it to a slice), followed by capacity-len logically uninitialized, contiguous elements.

I didn't use `MaybeUninit` because 'm not sure about the compiler optimization and I'm lazy to try it out XD and now you caught it. I will try it later and if the optimization is fine I will use this.

Reply:
I rewrote use `MaybeUninit`. Here is a short demo (see `batch_get_2`). Do you think this is what you want?

```rust
pub fn batch_get_1(keys: &[Vec<u8>]) -> Vec<Vec<u8>> {
    let order_and_keys = keys.iter().enumerate();
    let mut values = Vec::with_capacity(keys.len());
    unsafe {
        for (original_order, key) in order_and_keys {
            let value = get(key);
            mem::forget(mem::replace(
                values.get_unchecked_mut(original_order),
                value,
            ));
        }
        values.set_len(keys.len());
    }
    values
}

pub fn batch_get_2(keys: &[Vec<u8>]) -> Vec<Vec<u8>> {
    let order_and_keys = keys.iter().enumerate();
    let mut values = Vec::with_capacity(keys.len());
    for _ in 0..keys.len() {
        values.push(mem::MaybeUninit::uninit());
    }
    for (original_order, key) in order_and_keys {
        let value = get(key);
        mem::replace(
            unsafe { values.get_unchecked_mut(original_order) },
            mem::MaybeUninit::new(value),
        );
    }

    let ret = unsafe { Vec::from_raw_parts(
        values.as_mut_ptr() as *mut Vec<u8>,
        values.len(),
        values.capacity(),
    ) };
    core::mem::forget(values);
    ret
}
```

@sticnarf

Reply:
```rust
pub fn batch_get(keys: &[Vec<u8>]) -> Vec<Vec<u8>> {
    let order_and_keys = keys.iter().enumerate();
    let mut values: Vec<MaybeUninit<Vec<u8>>> = Vec::with_capacity(keys.len());
    for _ in 0..keys.len() {
        values.push(mem::MaybeUninit::uninit());
    }
    for (original_order, key) in order_and_keys {
        let value = get(key);
        unsafe {
            values[original_order].as_mut_ptr().write(value);
        }
    }

    unsafe { mem::transmute::<Vec<MaybeUninit<Vec<u8>>>, Vec<Vec<u8>>>(values) }
}
```
I think code like this should be OK?